manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+transformation is a property of individual functions, not whole
+recursive groups.
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+  append_flatten xs xss
+
+and append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+
+| append_flatten xs xss
+  ^^^^^^^^^^^^^^^^^^^^^
+Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+but the function called is not itself specialized for TMC,
+so the call will not be transformed into a tail call.
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+
+Again the fix is to specialized "append_flatten" as well:
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+  append_flatten xs xss
+
+and[@tail_mod_cons] append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+\end{verbatim}
+
+Non-recursive functions can also be marked "[\@tail_mod_cons]"; this is
+typically useful for local bindings to recursive functions.
+
+Incorrect version:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+                           ^^^^^^^^^^^^^^^^^^^^^
+Warning 51 [wrong-tailcall-expectation]: expected tailcal
+\end{verbatim}
+
+Recommend fix:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let[@tail_mod_cons] self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{verbatim}
+
+In other cases, there is no benefit in making the called function
+tail-mod-cons, or it is not possible, for example it is a function
+parameter (the transformation only works with calls to
+known functions).
+
+For example, consider a substitution function on binary trees:
+\begin{verbatim}
+type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree
+
+let[@tail_mod_cons] rec bind (f : 'a -> 'a tree) (t : 'a tree) : 'a tree =
+  match t with
+  | Leaf v -> f v
+  | Node (left, right) -> Node (bind f left, (bind[@tailcall]) f right)
+
+|             f v
+              ^^^
+Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+but the function called is not itself specialized for TMC,
+so the call will not be transformed into a tail call.
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+
+Here "f" is a function parameter, not a direct call, and the current
+implementation is strictly first-order, it does not support
```

Comment:
Ah, maybe that's the terminology - f is a function value or even f is higher order?!?!

Reply:
`bind` is higher-order and `f v` is not a direct call / a call to a known function. (I also used "function parameter" which here somehow means both "a parameter that is a function" and the more common "a parameter of the function".)